sgRNA_design/sgRNA_finder.py

```diff
@@ -0,0 +1,98 @@
+from collections import defaultdict
+
+# Function to determine if a potential PAM site meets the specified range and 
+# efficiency values. pamLine should be one line of csv with format:
+# Start Pos,Stop Pos, sgRNA sequence, PAM sequence (NGG), strand, MIT score
+def pam_In_Range(pamLine,geneRange,effCutOff=0):
+
+	# Parse and split csv line into list of string
+	words = pamLine.strip("\n").strip(",").split(',')
+	# Calculate the end position for PAM sequence based on desired coverage
+	endRange = (geneRange[1] - geneRange[0])*geneRange[2] + geneRange[0]
+
+	# Can be called directly in if and return statements, but makes code
+	# difficult to read
+	pamSta = int(words[0]) # Start Position
+	pamStp = int(words[1]) # Stop Position
+	pamSeq = words[2]      # sgRNA sequence
+	pamNGG = words[3]      # PAM sequence
+	pamStr = int(words[4]) # plus or minus strand
+	pamEff = float(words[5]) # MIT efficiency score
+
+	# determine if the PAM falls within the desired start and end sequence, if
+	# it meets the desired efficiency cut off (set to 0 by default) and whether
+	# it has masked characters in the sequence. Returns any PAMs that pass as 
+	# a list of values. Returns empty string if it fails because None value
+	# causes error in generating dictionary from returned values
+	if pamSta > geneRange[0] and pamStp < endRange and 'N' not in pamSeq and \
+	pamEff > effCutOff:
+		return [pamSta,pamStp,pamSeq,pamNGG,pamStr,pamEff];
+	else: return "";
+
+# function to return dictionary with protein name as key and start & end
+# positions as values. Takes csv file with format:
+# Protein Name,Start Pos,End Pos,Coverage
+def protein_Sites(fn):
+	
+	with open(fn, "r") as f:
+		proteins = {}
+		for line in f:
+			# csv line formatting
+			words = line.strip("\n").strip(",").split(',')
+			proteins[words[0]] = (int(words[1]),int(words[2]),float(words[3]))
+	f.close()
+	return proteins;
+
+# Sort list of PAMs (which are list themselves) into ordered list based on start
+# position. Starting with first PAM, determines if two sites overlap (with 
+# extension for Cas9 size) and discards the lower scoring one. Is greedy, and 
+# will discard two non-overlapping PAMs for a third that overlaps both if that
+# one is higher scoring.
+def pam_Overlap(listOfPams,overlapExtension=0):
+
+	# I have no idea how this works, but it gives a sorted list of PAMs. Thank
+	# you StackOverflow
+	# stackoverflow.com/questions/5201191/sorting-a-list-of-lists-in-python
+	listOfPams.sort(key=lambda x: x[1])	
```

Comment:
Not sure if you figured it out but .... First you construct a temporary function (lambda function), which takes a value and returns a value used for sorting. For example, lets say you wanted to sort a set of Person Objects by name. You might do something like: `lambda PersonObj : PersonObj.name`. This function is set as the 'key' value parameter in the function sort. This is a parameter that the sort function uses to evaluate the 'value' of a particular index in the list. In your case, you choose to evaluate the 'value' of each index in your list by the first element of that object (which in your case, is actually another list!). Hence you sort by the first index of the list. That's how it works :)